### PR TITLE
Open Policy Agent (OPA) Validation Support

### PR DESCRIPTION
**Pull request by renuka-fernando**
### Purpose
$subject

### Issues

Fixes https://github.com/wso2/product-microgateway/issues/2678

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java [hunk 1]
```diff
@@ -32,51 +39,66 @@
...
+            }
+            case "OPA": {
+                return opaAuthValidation(requestContext, policy.getParameters());
             }
         }
+
+        // TODO: check placeholders in logs are working
+        log.error("Operation policy action \"{}\" is not supported", policy.getAction(),
+                ErrorDetails.errorLog(LoggingConstants.Severity.MINOR, 6100));
+        return false;
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:101]: return 500 to user

Comment (AmaliMatharaarachchi) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:104]: If the provided action is not matched with any of the policies why should we break the request? if any of the policies is not supported it should be validated in the API deploying time not here imho.

Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:104]: +1 The adapter validates policies against policy spec and definition when deploying APIs, but there could be cases where spec and definition may contain invalid actions. We can introduce these action names to Adapter too, so the Adapter also knows what are the supported actions.
We can log as a major (This error is due to our code issue - In adapter, we have not validated the action) error and continue the request flow. @praminda wdyt?

Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:111]: The adapter validation will be fixed with https://github.com/wso2/product-microgateway/issues/2692

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationConstants.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package org.wso2.choreo.connect.enforcer.interceptor;
+
+/**
+ * Common set of constants related to the Mediation Policies.
+ */
+public class MediationConstants {
+    // TODO: (renuka) check error codes with APIM
+    public static final int GENERAL_ERROR = 901100;
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationConstants.java:26]: Had the same concern. btw
- We maintain throttle related error codes https://github.com/wso2/product-microgateway/blob/main/enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/throttle/ThrottleConstants.java#L25
- API security: https://github.com/wso2/product-microgateway/blob/main/enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/constants/APISecurityConstants.java#L26

any suggestions for improvement?

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java [hunk 3]
```diff
@@ -130,4 +162,24 @@ private void modifyMethod(RequestContext requestContext, Map<String, String> pol
...
+    private boolean opaAuthValidation(RequestContext requestContext, Map<String, String> policyAttrib) {
+        try {
+            boolean isValid = OPAClient.getInstance().validateRequest(requestContext, policyAttrib);
+            if (!isValid) {
+                log.error("OPA validation failed for the request: " + requestContext.getRequestPath(),
+                        ErrorDetails.errorLog(LoggingConstants.Severity.MINOR, 6101));
+                FilterUtils.setErrorToContext(requestContext,
+                        APISecurityConstants.REMOTE_AUTHORIZATION_AUTH_FORBIDDEN,
+                        APIConstants.StatusCodes.UNAUTHORIZED.getCode(),
+                        "Request authorization failure at the remote authorization server");
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:175]: He may not aware about OPA mediation, so we generalized it to

HTTP 403
```json
{"code": 901101, "error_message": "Request authorization failure", "error_description": "Request authorization failure at the remote authorization server"}
```

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java [hunk 4]
```diff
@@ -32,51 +39,76 @@
...
+                    removeAllQueries(requestContext, policy.getParameters());
+                    return true;
+                }
+                case "REWRITE_RESOURCE_METHOD": {
+                    modifyMethod(requestContext, policy.getParameters());
+                    return true;
+                }
+                case "OPA": {
+                    return opaAuthValidation(requestContext, policy.getParameters());
+                }
```
Comment (slahirucd7) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:95]: Just to clarify: 
Can we add a default case (to handle unsupported actions apart from the null pointer situations) ? Or are we going to handle that scenario after line number 107 ?

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPADefaultRequestGenerator.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        // additional fields from choreo connect
+        inputPayload.put("apiName", requestContext.getMatchedAPI().getName());
+        inputPayload.put("apiVersion", requestContext.getMatchedAPI().getVersion());
+        inputPayload.put("orgId", requestContext.getMatchedAPI().getOrganizationId());
+        inputPayload.put("vhost", requestContext.getMatchedAPI().getVhost());
+        inputPayload.put("pathTemplate", requestContext.getRequestPathTemplate());
+        inputPayload.put("prodClusterName", requestContext.getProdClusterHeader());
+        inputPayload.put("sandClusterName", requestContext.getSandClusterHeader());
+        inputPayload.put("requestBody", requestContext.getRequestPayload());
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPADefaultRequestGenerator.java:47]: Do not want request body, can have a custom one if someone wants

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPAClient.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
+package org.wso2.choreo.connect.enforcer.interceptor.opa;
```
Comment (pubudu538) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPAClient.java:1]: License?

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationConstants.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationConstants.java]: moved to `GeneralErrorCodeConstants.java` with https://github.com/wso2/product-microgateway/pull/2688/commits/82e1e2bbdc60bbaa0a85f1d866415288fa23ee5f

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java]: removed description with https://github.com/wso2/product-microgateway/commit/82e1e2bbdc60bbaa0a85f1d866415288fa23ee5f

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPADefaultRequestGenerator.java (hunk 5 above)
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPADefaultRequestGenerator.java]: removed with https://github.com/wso2/product-microgateway/commit/82e1e2bbdc60bbaa0a85f1d866415288fa23ee5f

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java (hunk 4 above)
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:96]: yes this is handled in line 107

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPAClient.java (hunk 6 above)
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPAClient.java:19]: added with https://github.com/wso2/product-microgateway/commit/82e1e2bbdc60bbaa0a85f1d866415288fa23ee5f

**Review by AmaliMatharaarachchi — APPROVED** [review]

File: enforcer-parent/commons/src/main/java/org/wso2/choreo/connect/enforcer/commons/model/RequestContext.java [hunk 7]
```diff
@@ -172,6 +173,15 @@ public String getClientIp() {
         return clientIp;
     }
 
+    /**
+     * Returns HTTP Protocol of the HTTP request
```
Comment (praminda) [inline enforcer-parent/commons/src/main/java/org/wso2/choreo/connect/enforcer/commons/model/RequestContext.java:177]:
```suggestion
     * Returns HTTP Protocol of the HTTP request.
```

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPAClient.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        String serverUrl = policyAttrib.get("serverUrl");
+        String token = policyAttrib.get("accessToken");
+        String policyName = policyAttrib.get("policy");
+        String ruleName = policyAttrib.get("rule");
+        // TODO: (renuka) handle additionalProperties, check with APIM
+
+        String requestBody = requestGenerator.generateRequest(policyName, ruleName, null, requestContext);
+        String evaluatingPolicyUrl = serverUrl + "/" + policyName + "/" + ruleName; // including multiple "/" is fine.
```
Comment (praminda) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/opa/OPAClient.java:91]:
```suggestion
        String evaluatingPolicyUrl = serverUrl + '/' + policyName + '/' + ruleName; // including multiple "/" is fine.
```

**Review by pubudu538 — APPROVED** [review]

**Review by slahirucd7 — APPROVED** [review]

**Review by praminda — APPROVED** [review]

**Pull request merged**